omod/src/main/java/org/openmrs/module/queue/web/QueueEntryTransitionRestController.java

```diff
@@ -34,80 +36,82 @@
 		
 		// Transition Date
 		Date transitionDate = new Date();
-		if (body.containsKey(TRANSITION_DATE)) {
-			transitionDate = (Date) ConversionUtil.convert(body.get(TRANSITION_DATE), Date.class);
+		if (body.getTransitionDate() != null) {
+			transitionDate = (Date) ConversionUtil.convert(body.getTransitionDate(), Date.class);
 		}
 		if (transitionDate == null) {
-			throw new APIException("Invalid transition date specified: " + body.get(TRANSITION_DATE));
+			throw new APIException("Invalid transition date specified: " + body.getTransitionDate());
 		}
 		transition.setTransitionDate(transitionDate);
 		
 		// Queue
-		if (body.containsKey(NEW_QUEUE)) {
-			Optional<Queue> queueOptional = services.getQueueService().getQueueByUuid(body.get(NEW_QUEUE));
+		if (body.getNewQueue() != null) {
+			Optional<Queue> queueOptional = services.getQueueService().getQueueByUuid(body.getNewQueue());
 			if (!queueOptional.isPresent()) {
-				throw new APIException("Invalid queue specified: " + body.get(NEW_QUEUE));
+				throw new APIException("Invalid queue specified: " + body.getNewQueue());
 			}
 			transition.setNewQueue(queueOptional.get());
 		}
 		
 		// Status
-		if (body.containsKey(NEW_STATUS)) {
-			Concept concept = services.getConcept(body.get(NEW_STATUS));
+		if (body.getNewStatus() != null) {
+			Concept concept = services.getConcept(body.getNewStatus());
 			if (concept == null) {
-				throw new APIException("Invalid status specified: " + body.get(NEW_STATUS));
+				throw new APIException("Invalid status specified: " + body.getNewStatus());
 			}
 			transition.setNewStatus(concept);
 		}
 		
 		// Priority
-		if (body.containsKey(NEW_PRIORITY)) {
-			Concept concept = services.getConcept(body.get(NEW_PRIORITY));
+		if (body.getNewPriority() != null) {
+			Concept concept = services.getConcept(body.getNewPriority());
 			if (concept == null) {
-				throw new APIException("Invalid priority specified: " + body.get(NEW_PRIORITY));
+				throw new APIException("Invalid priority specified: " + body.getNewPriority());
 			}
 			transition.setNewPriority(concept);
 		}
 		
-		transition.setNewPriorityComment(body.get(NEW_PRIORITY_COMMENT));
+		transition.setNewPriorityComment(body.getNewPriorityComment());
 		
 		// Execute transition
 		QueueEntry newQueueEntry = services.getQueueEntryService().transitionQueueEntry(transition);
 		return ConversionUtil.convertToRepresentation(newQueueEntry, Representation.REF);
 	}
+	
+	@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/queue-entry/undo-transition", method = {
+	        RequestMethod.PUT, RequestMethod.POST })
+	@ResponseBody
+	public Object undoTransition(@RequestBody UndoQueueEntryTransitionRequest body) {
+		QueueEntryService qes = services.getQueueEntryService();
+		Optional<QueueEntry> queueEntry = qes.getQueueEntryByUuid(body.getQueueEntry());
+		if (queueEntry.isPresent()) {
+			QueueEntry unEndedQueueEntry = services.getQueueEntryService().undoTransition(queueEntry.get());
```

Comment:
Per above comments about throwing Exceptions, we should present these exceptions back to the user with appropriate error messages that can be adequately consumed in the UI